File: crates/normalize-languages/tests/fixtures/erlang/sample.erl
-module(sample).

-import(lists, [foldl/3, filter/2]).

-export([classify/1, sum_evens/1, factorial/1, member/2]).

-record(point, {x :: integer(), y :: integer()}).

-type coordinate() :: {integer(), integer()}.

-spec classify(integer()) -> negative | zero | positive.
%% Classify a number as negative, zero, or positive
classify(N) when N < 0 ->
    negative;
classify(0) ->
    zero;
classify(_N) ->
    positive.

%% Sum the even numbers in a list
sum_evens(List) ->
    Evens = filter(fun(X) -> X rem 2 =:= 0 end, List),
    foldl(fun(X, Acc) -> X + Acc end, 0, Evens).

%% Compute factorial recursively
factorial(0) ->
    1;
factorial(N) when N > 0 ->
    N * factorial(N - 1).

%% Check if an element is in a list
member(_Elem, []) ->
    false;
member(Elem, [Elem | _Rest]) ->
    true;
member(Elem, [_ | Rest]) ->
    member(Elem, Rest).

%% Create a point record
make_point(X, Y) ->
    #point{x = X, y = Y}.

File: crates/normalize-languages/tests/fixtures/ocaml/sample.ml
open List
open Printf

(* Module for stack operations *)
module Stack = struct
  type 'a t = { mutable items : 'a list }

  let create () = { items = [] }

  let push s x = s.items <- x :: s.items

  let pop s =
    match s.items with
    | [] -> None
    | x :: rest ->
        s.items <- rest;
        Some x

  let is_empty s = s.items = []
end

(* Type definition for a binary tree *)
type 'a tree =
  | Leaf
  | Node of 'a * 'a tree * 'a tree

(* Insert into BST *)
let rec insert x = function
  | Leaf -> Node (x, Leaf, Leaf)
  | Node (y, left, right) ->
      if x < y then Node (y, insert x left, right)
      else if x > y then Node (y, left, insert x right)
      else Node (y, left, right)

(** Classify a number as negative, zero, or positive. *)
[@inline]
let classify n =
  if n < 0 then "negative"
  else if n = 0 then "zero"
  else "positive"

(* Sum of even numbers in a list *)
let sum_evens lst =
  fold_left (fun acc x -> if x mod 2 = 0 then acc + x else acc) 0 lst

let () =
  let s = Stack.create () in
  Stack.push s 1;
  Stack.push s 2;
  printf "%s\n" (classify 5);
  printf "%d\n" (sum_evens [1; 2; 3; 4; 5])
